File: frontend/src/utils/contextUpdates.js
/** Client-side mirror of backend context_updates deduplication. */

import { formatDateTimeIST, parseApiDate } from './datetime';

/** Default number of timeline rows shown before "Load older activity". */
export const TIMELINE_INITIAL_VISIBLE = 15;

/** How many additional rows each load-more click reveals. */
export const TIMELINE_LOAD_MORE_STEP = 15;

export function normalizeDescription(text) {
  return (text || '').trim().toLowerCase().replace(/\s+/g, ' ');
}

function entryTimestamp(update) {
  const raw = update.timestamp_dt || update.timestamp;
  if (!raw) return 0;
  const d = parseApiDate(raw);
  return d ? d.getTime() : 0;
}

function dedupeKey(update) {
  const noteId = (update.note_id || update.external_id || '').trim();
  if (noteId) return `id:${noteId}`;
  const agent = (update.agent || '').trim().toLowerCase();
  const type = (update.type || 'note').trim().toLowerCase();
  const desc = normalizeDescription(update.description);
  return `content:${type}|${agent}|${desc}`;
}

export function dedupeContextUpdates(updates) {
  if (!updates?.length) return [];
  const sorted = [...updates].sort((a, b) => entryTimestamp(b) - entryTimestamp(a));
  const seen = new Set();
  const kept = [];
  for (const entry of sorted) {
    const key = dedupeKey(entry);
    if (seen.has(key)) continue;
    seen.add(key);
    kept.push(entry);
  }
  return kept.sort((a, b) => entryTimestamp(b) - entryTimestamp(a));
}

/**
 * Timeline entries for UI display: deduped, newest-first.
 * Do not reverse this list — index 0 is the latest activity.
 */
export function getTimelineForDisplay(updates) {
  return dedupeContextUpdates(updates);
}

export function formatTimelineAttribution(update) {
  const name = (update?.actor_name || update?.agent || 'Unknown').trim() || 'Unknown';
  const raw = update?.timestamp_dt || update?.timestamp;
  let timeStr = '';
  if (raw) {
    timeStr = formatDateTimeIST(raw) || String(raw);
  }
  return { name, timeStr, label: timeStr ? `Added by ${name} at ${timeStr}` : `Added by ${name}` };
}

export function contextUpdateKey(update) {
  const noteId = (update.note_id || update.external_id || '').trim();
  if (noteId) return `note-${noteId}`;
  const ts = update.timestamp || update.timestamp_dt || '';
  return `${update.type || 'note'}|${(update.agent || '').toLowerCase()}|${normalizeDescription(update.description)}|${ts}`;
}

File: backend/crm/services/context_updates.py
"""Deduplicate lead context_updates timeline entries."""
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any, Dict, List, Tuple

from crm.utils.helpers import coerce_datetime


def normalize_description(text: str) -> str:
    """Normalize note text for duplicate comparison."""
    text = (text or "").lower()
    text = re.sub(r"\[\d{4}-\d{2}-\d{2}\]", " ", text)
    text = re.sub(r"[^\w\s]", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def _entry_timestamp(entry: Dict[str, Any]) -> datetime:
    dt = entry.get("timestamp_dt")
    if isinstance(dt, datetime):
        t = dt
    else:
        t = coerce_datetime(entry.get("timestamp")) if entry.get("timestamp") else None
    if t is None:
        return datetime.min.replace(tzinfo=timezone.utc)
    if t.tzinfo is None:
        t = t.replace(tzinfo=timezone.utc)
    return t


def _dedupe_key(entry: Dict[str, Any]) -> Tuple:
    note_id = (entry.get("note_id") or entry.get("external_id") or "").strip()
    if note_id:
        return ("id", note_id)
    desc = normalize_description(entry.get("description") or "")
    etype = (entry.get("type") or "").strip().lower()
    agent = (entry.get("agent") or entry.get("actor_name") or "").strip().lower()
    return ("content", etype, agent, desc)


def dedupe_context_updates(updates: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Collapse duplicate timeline rows; keep the entry with the newest timestamp.

    Duplicates match on note_id when present, otherwise (type, agent, normalized description).
    """
    if not updates:
        return []

    sorted_newest_first = sorted(updates, key=_entry_timestamp, reverse=True)
    seen: set[Tuple] = set()
    kept: List[Dict[str, Any]] = []

    for entry in sorted_newest_first:
        key = _dedupe_key(entry)
        if key in seen:
            continue
        seen.add(key)
        kept.append(entry)

    kept.sort(key=_entry_timestamp, reverse=True)
    return kept
